programs/conditional_vault/src/instructions/redeem_tokens.rs

```diff
@@ -0,0 +1,127 @@
+use super::*;
+
+impl<'info, 'c: 'info> InteractWithVault<'info> {
+    pub fn validate_redeem_tokens(&self) -> Result<()> {
+        require!(
+            self.question.is_resolved(),
+            VaultError::CantRedeemConditionalTokens
+        );
+
+        Ok(())
+    }
+
+    pub fn handle_redeem_tokens(ctx: Context<'_, '_, 'c, 'info, Self>) -> Result<()> {
+        let accs = &ctx.accounts;
+
+        let (mut conditional_token_mints, mut user_conditional_token_accounts) =
+            Self::get_mints_and_user_token_accounts(&ctx)?;
+
+        // calculate the expected future supplies of the conditional token mints
+        // as current supply - user balance
+        let expected_future_supplies: Vec<u64> = conditional_token_mints
+            .iter()
+            .zip(user_conditional_token_accounts.iter())
+            .map(|(mint, account)| mint.supply - account.amount)
+            .collect();
+
+        let vault = &accs.vault;
+        let question = &accs.question;
+
+        let seeds = generate_vault_seeds!(vault);
+        let signer = &[&seeds[..]];
+
+        let user_underlying_balance_before = accs.user_underlying_token_account.amount;
+        let vault_underlying_balance_before = accs.vault_underlying_token_account.amount;
+        // safe because there is always at least two conditional tokens and thus
+        // at least two user conditional token accounts
+        let max_redeemable = user_conditional_token_accounts
+            .iter()
+            .map(|account| account.amount)
+            .max()
+            .unwrap();
+
+        let mut total_redeemable = 0;
+
+        for (conditional_mint, user_conditional_token_account) in conditional_token_mints
+            .iter()
+            .zip(user_conditional_token_accounts.iter())
+        {
+            // this is safe because we check that every conditional mint is a part of the vault
+            let payout_index = vault
+                .conditional_token_mints
+                .iter()
+                .position(|mint| mint == &conditional_mint.key())
+                .unwrap();
+
+            total_redeemable += ((user_conditional_token_account.amount as u128
+                * question.payout_numerators[payout_index] as u128)
+                / question.payout_denominator as u128) as u64;
```

Comment:
doing a mix of Solana `Context` and (unsafe) math logic within one function. would be great to split up math and state modification logic from solana account loader stuff. this makes rust tests much easier to write and lets you fuzz more easily